docs/proposals/nodeselector.md

```diff
@@ -0,0 +1,174 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/nodeselector.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NodeSelector, NodeConstraints, and NodePreferences
+
+## Introduction
+
+This document proposes a new label selector representation, called `NodeSelector`,
+that is similar in many ways to `PodSelector`, but is a bit more flexible and is
+intended to be used for selecting nodes rather than pods.
+
+In addition, we propose to replace the `map[string]string` in `PodSpec` that the scheduler
+currently uses as part of restricting the set of nodes onto which a pod is
+eligible to schedule, with a field called `NodeConstraints` that wil be represented
+by a `NodeSelector`. (In practice, as discussed later,
+we will *add* the `NodeConstraints` field rather than replacing `map[string]string`,
+due to backward compatibility requirements.) We further propose an additional
+new field for `PodSpec`, called `NodePreferences`, that specifies which nodes are
+preferred for scheduling among those that meet the hard requirements described
+by `NodeConstraints`.
+
```

Comment:
This approach favours a binary distinction between constraints (which are absolutely required) and preferences (which are implicitly all equally ignorable), and bakes this choice into the API structure.  I think I would prefer unifying these two concepts into one (e.g. `NodeAffinity`), where each such affinity is annotated with a relative strength.  Affinities annotated with the maximum possible relative strength (say 100%) are hard constraints, and all others are softer preferences with potentially different relative strengths (although in a first version you might only allow one non-maximul value, in which case you have the same semantics as proposed by NodeConstraints and NodePreferences, except that it's much more easily extensible by allowing other strength values in future.